File: Data Structures & Algorithms/coin-change/submission-16.py
class Solution:
    def coinChange(self, coins, amount):
        memo = [-1] * (amount + 1)
        def dfs(curr):
            if curr == amount:
                return 0

            if memo[curr] != -1:
                return memo[curr]
            
            res = float('inf')
            for c in coins:
                if curr + c <= amount:
                    res = min(res, 1 + dfs(curr + c))
            
            memo[curr] = res
            return res
        
        res = dfs(0)
        return -1 if res >= float('inf') else res

File: Data Structures & Algorithms/decode-ways/submission-16.java
class Solution {
    public int dfs(int i, String s, Map<Integer, Integer> memo) {
        if (memo.containsKey(i)) {
            return memo.get(i);
        }
        
        if (s.charAt(i) == '0') {
            return 0;
        }

        int res = dfs(i + 1, s, memo);

        if (i < s.length() - 1) {
            if (s.charAt(i) == '1' || s.charAt(i) == '2' && s.charAt(i + 1) < '7') {
                res += dfs(i + 2, s, memo);
            }
        }

        memo.put(i, res);
        return res;
    }
    public int numDecodings(String s) {
        Map<Integer, Integer> memo = new HashMap<>();
        memo.put(s.length(), 1);
        return dfs(0, s, memo);
    }
}
